### PR TITLE
Delete project

### PR DESCRIPTION
**Pull request by ghost**
Each project now store the amount of credits that a user put into.
So when the project is deleted, every user who spend credits on it will have them back (The project owner you started the action will have his state automatically updated)

An array of custom type (userId, credits) is stored in DB.

### CODE REVIEW

File: api/lib/handlers/project/give/index.js [hunk 1]
```diff
@@ -37,19 +38,32 @@ module.exports.post = function (req, res, next) {
                     return next({ message: error.UNKNOWN_PROJECT });
                 else if (result.estimate - result.acquired < amount)
                     return next({ message: error.TOO_MUCH_CREDIT });
-                else if (resCheck.credit < amount)
+                else if (parseFloat(resCheck.credit) < amount)
```
Comment (prevostc) [inline api/lib/handlers/project/give/index.js:41]: can we make this parseFloat happen sooner ?

File: api/lib/handlers/project/give/index.js [hunk 2]
```diff
@@ -37,19 +38,32 @@ module.exports.post = function (req, res, next) {
...
+                We need to parse it to make it JS-usable
+                Once the prepared SQL statement has been made, we can just add it to the query
+            */
+            let i = 0;
+            const query = resProject.associate_users && postgresArray.parse(resProject.associate_users).find(user => {
+                user = eval(user.replace(/\(/g, '[').replace(/\)/g, ']'));
+                i++;
+                return user[0] === userId;
+            }) ? `associate_users[${i}].credits_amount = associate_users[${i}].credits_amount + ${amount}` : `associate_users = associate_users || '{"(${userId}, ${amount})"}'`;
+            DatabaseService(`UPDATE give_me_time_public.project SET acquired=acquired+($1), ${query} WHERE id=($2) RETURNING *`,
```
Comment (prevostc) [inline api/lib/handlers/project/give/index.js:66]: please add some tests

File: api/lib/handlers/project/index.js [hunk 3]
```diff
@@ -54,12 +54,33 @@ module.exports.delete = function (req, res, next) {
...
+            const parsedArray = postgresArray.parse(resProject.associate_users);
+            resProject.associate_users && parsedArray.map((user, index) => {
+                user = eval(user.replace(/\(/g, '[').replace(/\)/g, ']'));
+                DatabaseService('UPDATE give_me_time_public.person SET credit=credit+($1) WHERE id=($2) RETURNING *',
+                [user[1], user[0]], next,
+                result => {
+                    const newCredits = result.id === userId ? result.credit : null; // Store the project owner credits to refresh the state
+                    if (index + 1 === parsedArray.length)
+                        return cb(null, newCredits); // call cb when we went through all the array
+                });
```
Comment (prevostc) [inline api/lib/handlers/project/index.js:76]: please add some tests

File: api/lib/handlers/project/give/index.js (hunk 1 above)
Comment (ghost) [reply api/lib/handlers/project/give/index.js:41]: not really.. resCheck is the direct result of the 'checkUser' function.
This parseFloat is also only useful for the 'if' check as we don't use it anywhere else.

File: api/lib/handlers/project/give/index.js (hunk 2 above)
Comment (ghost) [reply api/lib/handlers/project/give/index.js:66]: Done

File: api/lib/handlers/project/index.js (hunk 3 above)
Comment (ghost) [reply api/lib/handlers/project/index.js:76]: Done

**Pull request merged**